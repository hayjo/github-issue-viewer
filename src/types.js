// @flow

type RootStackParamList = {
  Issue: 'undefined',
  Repository: 'undefined',
  Browser: { uri: string, title: string | 'undefined' },
};

export type { RootStackParamList };
