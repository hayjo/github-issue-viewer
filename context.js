import { createContext } from 'react';

const RepoContext = createContext();

export { RepoContext };
