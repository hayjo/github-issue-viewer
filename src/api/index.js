// @flow

import { mockSearchRepoResult, mockIssues } from './mockData';
import { MESSAGE, LIMIT } from '../constants';

const useMock = process.env.NODE_ENV === 'test';
const MOCK_DELAY = 500;

type Item = {
  id: number,
  name: string,
  owner: { id: number, login: string },
  description: string,
};

type TSearchRepoByQuery = Promise<{
  totalCount: number,
  items: Array<Item>,
}>;

async function searchRepoByQuery(
  query: string,
  page: number,
): TSearchRepoByQuery {
  if (useMock) {
    await new Promise(resolve => setTimeout(resolve, MOCK_DELAY));

    return new Promise(resolve => resolve(mockSearchRepoResult));
  }

  try {
    const response = await fetch(
      `https://api.github.com/search/repositories?q=${query}&sort=stars&order=desc&per_page=${LIMIT.SEARCH_REPO_PER_PAGE}&page=${page}`,
    );

    if (!response.ok) {
      throw new Error(MESSAGE.NETWORK_ERROR);
    }

    const {
      total_count: totalCount,
      incomplete_results: hasRateExceeded,
      items,
    } = await response.json();

    // https://docs.github.com/en/rest/reference/search#timeouts-and-incomplete-results
    if (hasRateExceeded) {
      throw new Error(MESSAGE.TEMPORARILY_UNAVAILABLE);
    }

    return { totalCount, items };
  } catch (err) {
    throw err;
  }
}

type Issue = {
  id: number,
  fullName: string,
  number: number,
  title: string,
  url: string,
  labels: Array<{ id: number, name: string, color: string }>,
  comments: number,
  createdAt: Date,
  assignees: Array<string>,
};

type TGetIssuesByRepoFullName = Promise<Issue[]>;

async function getIssuesByRepoFullName(
  fullName: string,
  page: number = 1,
): TGetIssuesByRepoFullName {
  if (useMock) {
    await new Promise(resolve => setTimeout(resolve, MOCK_DELAY));

    return new Promise(resolve =>
      resolve(
        mockIssues.map(issue => ({
          ...issue,
          url: issue.html_url,
          number: issue.number * page,
          fullName,
          createdAt: new Date(issue.created_at),
          assignees: issue.assignees.map(({ avatar_url }) => avatar_url),
        })),
      ),
    );
  }

  await new Promise(resolve => setTimeout(resolve, MOCK_DELAY));

  try {
    const response = await fetch(
      `https://api.github.com/repos/${fullName}/issues?per_page=100&page=${page}`,
    );

    if (!response.ok) {
      throw new Error(MESSAGE.NETWORK_ERROR);
    }

    const issueList = await response.json();
    const filteredIssueList = issueList.map(
      ({
        id,
        number,
        title,
        html_url,
        labels,
        comments,
        created_at,
        assignees,
      }) => ({
        id,
        fullName,
        number,
        title,
        url: html_url,
        labels,
        comments,
        createdAt: new Date(created_at),
        assignees: assignees.map(({ avatar_url }) => avatar_url),
      }),
    );

    return filteredIssueList;
  } catch (err) {
    throw err;
  }
}

export type { Issue };
export { searchRepoByQuery, getIssuesByRepoFullName };
