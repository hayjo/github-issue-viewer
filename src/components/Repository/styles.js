// @flow

import { StyleSheet } from 'react-native';
import type {
  ViewStyleProp,
  TextStyleProp,
  ImageStyleProp,
} from 'react-native/Libraries/StyleSheet/StyleSheet';

import { COLORS } from '../../constants';

const viewStyles: { [string]: ViewStyleProp } = StyleSheet.create({
  container: {
    backgroundColor: COLORS.DEFAULT,
    width: '100%',
    height: '100%',
  },
  cardContainer: {
    marginTop: 16,
  },
  card: {
    borderRadius: 2,
    borderWidth: 1,
    borderColor: COLORS.BORDER,
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingVertical: 10,
    paddingHorizontal: 8,
    marginVertical: 4,
    marginHorizontal: 8,
  },
  cardContent: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  cardTextContainer: {
    marginHorizontal: 12,
  },
});

const textStyles: { [string]: TextStyleProp } = StyleSheet.create({
  repoOwner: {
    color: COLORS.TEXT,
  },
  repoName: {
    marginTop: 4,
    fontSize: 16,
  },
});

const imageStyles: { [string]: ImageStyleProp } = StyleSheet.create({
  avatar: {
    height: 40,
    width: 40,
    borderRadius: 40,
    marginHorizontal: 4,
    marginVertical: 4,
  },
});

export { viewStyles, textStyles, imageStyles };
