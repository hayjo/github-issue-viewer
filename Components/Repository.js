// @flow

import React, { useContext, useState } from 'react';
import type { Node } from 'react';
import {
  SafeAreaView,
  Button,
  Image,
  Text,
  View,
  Alert,
  StyleSheet,
  Pressable,
} from 'react-native';
import type {
  ViewStyleProp,
  TextStyleProp,
  ImageStyleProp,
} from 'react-native/Libraries/StyleSheet/StyleSheet';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

import Search from '../Components/Search';

// import type { RootStackParamList } from '../types/nativeStack';
import { RepoContext } from '../context';
import { COLORS, SIZE } from '../constants';

Icon.loadFont();

function getProfileUrl(userId: string): string {
  return `https://avatars.githubusercontent.com/u/${userId}?v=4`;
}

const Repository: () => Node = () => {
  const { selectedRepoList, onSelectRepo } = useContext(RepoContext);
  const [showSearch, setShowSearch] = useState(false);

  const handleSelect = id => {
    const target = selectedRepoList.find(repo => repo.id === id);

    Alert.alert(
      'Delete',
      `${target.name} repository를 정말 삭제하시겠습니까?`,
      [
        {
          text: '삭제',
          onPress: () => onSelectRepo(target),
          style: 'cancel',
        },
        {
          text: '취소',
          onPress: () => {},
        },
      ],
    );
  };

  return (
    <SafeAreaView style={viewStyles.container}>
      {showSearch ? (
        <Search onCancel={() => setShowSearch(false)} />
      ) : (
        <View>
          <Button
            title="Search Repository"
            onPress={() => setShowSearch(true)}
          />
          <View style={viewStyles.cardContainer}>
            {selectedRepoList.map(({ id, name, ownerId, ownerName }) => (
              <Pressable
                key={id}
                style={viewStyles.card}
                onPress={() => handleSelect(id)}>
                <View style={viewStyles.cardContent}>
                  <Image
                    key={id}
                    source={{ uri: getProfileUrl(ownerId) }}
                    style={imageStyles.avatar}
                  />
                  <View style={viewStyles.cardTextContainer}>
                    <Text style={textStyles.repoOwner}>{ownerName}</Text>
                    <Text style={textStyles.repoName}>{name}</Text>
                  </View>
                </View>
                <Icon color={COLORS.BORDER} name="close" size={SIZE.ICON} />
              </Pressable>
            ))}
          </View>
        </View>
      )}
    </SafeAreaView>
  );
};

const viewStyles: { [string]: ViewStyleProp } = StyleSheet.create({
  container: {
    backgroundColor: COLORS.DEFAULT,
    width: '100%',
    height: '100%',
  },
  cardContainer: {
    marginTop: 16,
  },
  card: {
    borderRadius: 2,
    borderWidth: 1,
    borderColor: COLORS.BORDER,
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingVertical: 10,
    paddingHorizontal: 8,
    marginVertical: 4,
    marginHorizontal: 8,
  },
  cardContent: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  cardTextContainer: {
    marginHorizontal: 12,
  },
});

const textStyles: { [string]: TextStyleProp } = StyleSheet.create({
  repoOwner: {
    color: COLORS.TEXT,
  },
  repoName: {
    marginTop: 4,
    fontSize: 16,
  },
});

const imageStyles: { [string]: ImageStyleProp } = StyleSheet.create({
  avatar: {
    height: 40,
    width: 40,
    borderRadius: 40,
    marginHorizontal: 4,
    marginVertical: 4,
  },
});

export default Repository;
