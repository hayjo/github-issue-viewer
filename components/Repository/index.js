// @flow

import React, { useContext, useState } from 'react';
import type { Node } from 'react';
import {
  SafeAreaView,
  Button,
  Image,
  Text,
  View,
  Alert,
  Pressable,
} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

import Search from '../../components/Search';
import { RepoContext } from '../../context';
import { COLORS, SIZE } from '../../constants';
import { viewStyles, textStyles, imageStyles } from './styles';

Icon.loadFont();

function getProfileUrl(userId: string): string {
  return `https://avatars.githubusercontent.com/u/${userId}?v=4`;
}

const Repository: () => Node = () => {
  const { selectedRepoList, onSelectRepo } = useContext(RepoContext);
  const [showSearch, setShowSearch] = useState(false);

  const handleSelect = id => {
    const target = selectedRepoList.find(repo => repo.id === id);

    Alert.alert(
      'Delete',
      `${target.name} repository를 정말 삭제하시겠습니까?`,
      [
        {
          text: '삭제',
          onPress: () => onSelectRepo(target),
          style: 'cancel',
        },
        {
          text: '취소',
          onPress: () => {},
        },
      ],
    );
  };

  return (
    <SafeAreaView style={viewStyles.container}>
      {showSearch ? (
        <Search onCancel={() => setShowSearch(false)} />
      ) : (
        <View>
          <Button
            title="Search Repository"
            onPress={() => setShowSearch(true)}
          />
          <View style={viewStyles.cardContainer}>
            {selectedRepoList.map(({ id, name, ownerId, ownerName }) => (
              <Pressable
                key={id}
                style={viewStyles.card}
                onPress={() => handleSelect(id)}>
                <View style={viewStyles.cardContent}>
                  <Image
                    key={id}
                    source={{ uri: getProfileUrl(ownerId) }}
                    style={imageStyles.avatar}
                  />
                  <View style={viewStyles.cardTextContainer}>
                    <Text style={textStyles.repoOwner}>{ownerName}</Text>
                    <Text style={textStyles.repoName}>{name}</Text>
                  </View>
                </View>
                <Icon color={COLORS.BORDER} name="close" size={SIZE.ICON} />
              </Pressable>
            ))}
          </View>
        </View>
      )}
    </SafeAreaView>
  );
};

export default Repository;
