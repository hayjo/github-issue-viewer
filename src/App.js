// @flow

import React, { useState, useEffect } from 'react';
import type { Node } from 'react';
import { Text } from 'react-native';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { enableScreens } from 'react-native-screens';
import Icon from 'react-native-vector-icons/Octicons';

import useStoredList from './hooks/useStoredList';
import { RepoContext } from './context';
import IssueViewer from './components/IssueViewer';
import Repository from './components/Repository';
import Browser from './components/common/Browser';
import Error from './components/Error';
import { LIMIT, SIZE } from './constants';
import type { RootStackParamList } from './types';

Icon.loadFont();
enableScreens(true);

const Stack = createNativeStackNavigator<RootStackParamList>();

const App: () => Node = () => {
  const {
    list: repoList,
    updateList,
    hasExceedLimit,
    confirmLimitNotice,
    error,
  } = useStoredList({
    key: 'repoList',
    limit: LIMIT.REPO_COUNT,
  });
  const [hasError, setHasError] = useState(false);

  const notifyError = err => {
    console.log(err);
    setHasError(true);
  };

  useEffect(() => {
    if (error) {
      setHasError(true);
    }
  }, [error]);

  return (
    <RepoContext.Provider
      value={{
        selectedRepoList: repoList,
        hasExceedLimit,
        onSelectRepo: updateList,
        onNoticeClick: confirmLimitNotice,
        notifyError,
      }}>
      {hasError ? (
        <Error />
      ) : (
        <NavigationContainer>
          <Stack.Navigator initialRouteName="Issue">
            <Stack.Screen
              name="Issue"
              component={IssueViewer}
              options={({ navigation }) => ({
                headerRight: () => (
                  <Text onPress={() => navigation.navigate('Repository')}>
                    <Icon name="repo" size={SIZE.ICON} />
                  </Text>
                ),
              })}
            />
            <Stack.Screen name="Repository" component={Repository} />
            <Stack.Screen
              name="Browser"
              component={Browser}
              options={({ route }) => ({ title: route.params.title })}
            />
          </Stack.Navigator>
        </NavigationContainer>
      )}
    </RepoContext.Provider>
  );
};

export default App;
