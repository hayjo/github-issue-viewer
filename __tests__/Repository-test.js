import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import Repository from '../src/components/Repository';
import { RepoContext } from '../src/context';

const initialData = {
  selectedRepoList: [],
  hasExceedLimit: false,
  onSelectRepo: jest.fn(),
  onNoticeClick: jest.fn(),
  notifyError: jest.fn(),
};

describe('Repository component test', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <RepoContext.Provider value={initialData}>
          <Repository onCancel={jest.fn()} />
        </RepoContext.Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
