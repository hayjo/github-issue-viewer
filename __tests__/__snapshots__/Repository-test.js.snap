// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository component test renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <View>
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Search Repository
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "marginTop": 16,
        }
      }
    />
  </View>
</RCTSafeAreaView>
`;
